core/src/services/hdfs/writer.rs

```diff
@@ -44,18 +78,42 @@ impl oio::Write for HdfsWriter<hdrs::AsyncFile> {
             .map_err(new_std_io_error)
     }
 
+    fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<Result<()>> {
+        loop {
+            if let Some(mut fut) = self.fut.take() {
+                let res = ready!(fut.poll_unpin(cx));
+                return Poll::Ready(res);
+            }
+
+            let _ = Pin::new(&mut self.f)
+                .poll_close(cx)
```

Comment:
We can follow the same pattern in fs services:

https://github.com/apache/incubator-opendal/blob/48d88671e7e9e658d14374d42af530f0a4b9e654/core/src/services/fs/writer.rs#L67-L91